utils/wasm-instrument/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+        // {
+        Instruction::GetLocal(0),
+        Instruction::I64ExtendUI32,
+        Instruction::I64Const(i64::MAX),
+        Instruction::I64Add,
+        // }
+        // gas -= (gas_to_charge + cost_for_func)
+        // {
+        Instruction::I64Sub,
+        Instruction::SetGlobal(gas_index),
+        // }
+        // check if there is enough gas allowance
+        Instruction::GetGlobal(allowance_index),
+        // calculate gas_to_charge + cost_for_func
+        // {
+        Instruction::GetLocal(0),
+        Instruction::I64ExtendUI32,
+        Instruction::I64Const(i64::MAX),
+        Instruction::I64Add,
+        // }
+        // if allowance < (gas_to_charge + cost_for_func)
+        Instruction::I64LtU,
+        Instruction::If(elements::BlockType::NoResult),
+        Instruction::Call(out_of_allowance_index),
+        Instruction::Unreachable,
+        Instruction::End,
+        // update gas allowance
+        Instruction::GetGlobal(allowance_index),
+        // calculate gas_to_charge + cost_for_func
+        // {
+        Instruction::GetLocal(0),
+        Instruction::I64ExtendUI32,
+        Instruction::I64Const(i64::MAX),
+        Instruction::I64Add,
+        // }
+        // allowance -= (gas_to_charge + cost_for_func)
+        // {
+        Instruction::I64Sub,
+        Instruction::SetGlobal(allowance_index),
+        // }
+        Instruction::End,
+    ];
+
+    // determine cost for successful execution
+    let cost_check_blocks = match elements.iter().take(7).try_fold(0u64, |cost, instruction| {
+        rules.instruction_cost(instruction).map(|c| cost + c as u64)
+    }) {
+        Some(c) => 2 * c,
+        None => return Err(mbuilder.build()),
+    };
+
+    let cost_set_blocks = match elements
+        .iter()
+        .skip(10)
+        .take(7)
+        .try_fold(0u64, |cost, instruction| {
+            rules.instruction_cost(instruction).map(|c| cost + c as u64)
+        }) {
+        Some(c) => 2 * c,
+        None => return Err(mbuilder.build()),
+    };
+
+    let cost_call = match rules.instruction_cost(&Instruction::Call(0)) {
+        Some(c) => c as u64,
+        None => return Err(mbuilder.build()),
+    };
+
+    let cost = (cost_call + cost_check_blocks + cost_set_blocks) as i64;
+
+    // update cost for 'gas_charge' function itself
+    for instruction in elements
+        .iter_mut()
+        .filter(|i| matches!(i, Instruction::I64Const(_)))
+    {
+        *instruction = Instruction::I64Const(cost);
+    }
+
+    // gas_charge function
+    mbuilder.push_function(
+        builder::function()
+            .signature()
+            .with_param(ValueType::I32)
+            .build()
+            .body()
+            .with_instructions(elements::Instructions::new(elements))
+            .build()
+            .build(),
+    );
+
+    // back to plain module
+    let module = mbuilder.build();
+
+    gas_metering::post_injection_handler(module, rules, gas_charge_index, out_of_gas_index, 2)
```

Comment:
Add please comment about call indexes resolving here, because it's not simple to understand why call indexes stay correct after new imports push.

Reply:
Yep, good remark too. However there are comments in the original wasm-instrument explaining that:
- https://github.com/gear-tech/wasm-instrument/blob/gear-stable/src/gas_metering/mod.rs#L179
- https://github.com/gear-tech/wasm-instrument/blob/gear-stable/src/gas_metering/mod.rs#L187
- https://github.com/gear-tech/wasm-instrument/blob/gear-stable/src/gas_metering/mod.rs#L130

So I guess there is no reason to duplicate the information there.

Reply:
Ok